parlai/core/tod/tod_test_utils/test_agents.py

```diff
@@ -0,0 +1,214 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""
+Helpers so we don't need to create agents all over.
+"""
+
+import parlai.core.tod.tod_agents as tod_agents
+import parlai.core.tod.tod_core as tod_core
+
+import os
+
+API_DATABASE_FILE = os.path.join(
+    os.path.dirname(os.path.realpath(__file__)), "standalone_api_file.pickle"
+)
+
+
+def episode_has_broken_api_turn(episode_idx, max_turns):
+    """
+    To generate predictably broken episodes (for calculating metrics)
+    """
+    return episode_idx % 2 == 1 and max_turns > 0
+
+
+def turn_has_broken_api_call(round_idx, episode_idx):
+    """
+    To generate predictably broken turns (for calculating metrics)
+    """
+    return episode_idx % 2 == 1 and round_idx % 3 == 1
+
+
+def make_api_call_machine(round_idx, episode_idx=0, use_broken_mock_api_calls=False):
+    if round_idx == 0:
+        return {}
+    if use_broken_mock_api_calls:
+        # Hack as a way to test metrics reporting in tod world script
+        if turn_has_broken_api_call(round_idx, episode_idx):
+            round_idx = -1 * round_idx
+    return {tod_core.STANDARD_API_NAME_SLOT: f"name_{round_idx}", "in": round_idx}
+
+
+def make_api_resp_machine(round_idx):
+    if round_idx == 0:
+        return {}
+    return {"out": round_idx}
+
+
+def make_api_schemas_machine(max_rounds):
+    return [
+        {
+            tod_core.STANDARD_API_NAME_SLOT: f"name_{round_idx}",
+            tod_core.STANDARD_REQUIRED_KEY: ["in"],
+            tod_core.STANDARD_OPTIONAL_KEY: [],
+        }
+        for round_idx in range(1, max_rounds)
+    ]
+
+
+def make_goal_calls_machine(max_rounds):
+    return [make_api_call_machine(x) for x in range(1, max_rounds)]
+
+
+def get_rounds(episode_idx, max_rounds, use_broken_mock_api_calls=False):
+    return [
+        tod_core.TodStructuredRound(
+            user_utt=f"user_utt_{episode_idx}_{round_idx}",
+            api_call_machine=make_api_call_machine(
+                round_idx, episode_idx, use_broken_mock_api_calls
+            ),
+            api_resp_machine=make_api_resp_machine(round_idx),
+            sys_utt=f"sys_utt_{episode_idx}_{round_idx}",
+        )
+        for round_idx in range(max_rounds)
+    ]
+
+
+def get_round_utts(episode_idx, max_rounds, filter_utts=None):
+    if max_rounds < 1:
+        return []
+    utts = [
+        [
+            f"USER: user_utt_{episode_idx}_0",
+            "APICALL: ",
+            "APIRESP: ",
+            f"SYSTEM: sys_utt_{episode_idx}_0",
```

Comment:
why not use the tod_core constants for USER, APICALL, etc?

Reply:
Oof good minor catch.